Unbounded GC Thread Pool Expansion for CRIU Restore

RSalman opened the pull request:
Initial set of changes to expand the GC thread pool beyond the dispatcher initialization done at VM startup. These changes are required by the consumed OMR GC dispatcher/configuration APIs called for restore.

- Converted reference obj list iteration count from `gcThreadCount` to `regionExtension->_maxListIndex`, this is more correct. This is the only instance in the code where iteration implies the listCount from GC thread count. This inference was not an issue as the lists are inited based on the GC thread count, however now that the thread count may change, it is incorrect to assume the list count from the thread count.
- Release restore thread's VM access for the duration of `reinitializeGCThreadCountForRestore`
- Introduced `j9gc_reinitializeDefaults` in `mminit`, this is meant to be the top level init/verify method (e.g parsing/initing GC params based on restore cmd line opts). This is symmetrical restore call to VM startup `gcInitializeDefaults` call .
- Introduced `setMaxGCThreadCount` in Configuration Delegate to simply computation of default gc thread count at restore time

Signed-off-by: Salman Rana <salman.rana@ibm.com>

RSalman commented:
@amicic @dmitripivkine @tajila please review

amicic commented on runtime/gc_base/modronapi.cpp:
```diff
@@ -1123,7 +1125,11 @@ j9gc_reinitialize_for_restore(J9VMThread *vmThread)
 	MM_VerboseManagerBase *verboseGCManager = extensions->verboseGCManager;
 
+	/* The checkpoint thread must release VM access for the duration of reinitializeGCThreadCountForRestore.
+	 * The thread startup/attach may require exclusive access if an af/heap OOM is encountered.  */
```
2 things:
- simply state it may trigger GC (what is very similar to AF, but not really to heap OOM) - it's common knowledge that GC will need exclusive VM access
- need to state it'ss tartup of a new thread (rather then something about this thread triggering GC - in which case releasing VM access for this thread would be completely wrong).

So, something like this but feel free to change it further.

The checkpoint thread must release VM access for the duration of reinitializeGCThreadCountForRestore, since new GC threads could be started and the startup/attach of a new GC thread involves allocation and may trigger GC

amicic commented on runtime/gc_modron_startup/mmparseXXgc.cpp:
```diff
@@ -1530,6 +1530,16 @@ gcParseXXgcArguments(J9JavaVM *vm, char *optArg)
 		}
 
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+		if (try_scan(&scan_start, "RestoreGCThreads=")) {
```
hm, I'd rather not introduce a new option, but rely on existing ones
it's just that the option parsing logic would be extended to know if it's a restore (vs fresh VM start) so that value is stored into special restoreGCthreadCount variable.

tajila replied:
you can look at `vm->checkpointState.restoreArgsList` with something like `FIND_AND_CONSUME_ARG` to search for restore options

RSalman commented:
Latest changes:
- Introduced `j9gc_reinitializeDefaults` in `mminit`, this is meant to be the top level init/verify method (e.g setting params based on restore cmd line opts). This is symmetrical restore call to `gcInitializeDefaults` called during VM startup.
- Introduced `setMaxGCThreadCount` for Configuration Delegate to simply computation of default gc thread count at restore time, see related OMR for use.

amicic commented on runtime/gc_glue_java/ConfigurationDelegate.hpp:
```diff
@@ -354,6 +354,11 @@ class MM_ConfigurationDelegate
 
+	void setMaxGCThreadCount(MM_EnvironmentBase* env, uintptr_t maxGCThreads)
+	{
+		_maximumDefaultNumberOfGCThreads = maxGCThreads;
```
should this not be min(_maximumDefaultNumberOfGCThreads, maxGCThreads) - we might lose the hard limit?

amicic replied:
after looking elsewhere and realizing this is intentional to reduce the limit and the useage of this API is temporary, this is ok as-is

amicic commented:
Jenkins test sanity win jdk11

amicic commented:
Jenkins compile aix jdk8

amicic commented:
Jenkins test sanity xlinuxcriu jdk17

amicic commented:
Jenkins compile aix jdk8

amicic commented:
Restarted a failed aix compile, although this seemed like an infra problem. Merging...
